Implement generic injectable contributions

tortmayr opened the pull request:
- Generic theia-backend  contribution that can launch the modelserver if needed
- Frontend service for communicating with the modelserver
- Example application with a simple demo menu for sending requests to the model server

Note: This is the counterpart PR to https://github.com/eclipsesource/modelserver/pull/22. A running modelserver instance is required for testing.
The example application is already prepared for downloading and launching the latest modelserver snapshot autonomously. Once the counterpart PR for the modelserver is merged this can be activated.
(by setting the isRunning flag in the LaunchOptions to false, and activate the download-server script on build)

tortmayr committed:
Implement generic injectable contributions

- Generic theia-backend  contribution that can launch the modelserver if needed
- Frontend service for communicating with the modelserver
- Example application with a simple demo menu for sending requests to the model server

edgarmueller reviewed (commented):
I think the README is outdated as the `browser-app` (among other folders) has been moved. Could you update it?

tortmayr committed:
Update README.md

edgarmueller commented on .gitignore:
```diff
@@ -2,5 +2,7 @@ node_modules
+example/*-app/*
+!example/*-app/package.json
+
+example/modelsever-example/build/com\.eclipsesource\.modelserver\.example-0\.0\.1-SNAPSHOT-standalone\.jar
```
The folder name has a typo, it should read modelse**r**ver-example. The folder should be renamed accordingly.

tortmayr replied:
> Wrong folder name. If I follow 'Developing with the Browser example' section in the README I also run into issues with `Failed to resolve module: modelserver-example`. Probably related to the wrong folder name?

It works for me so I don't think this is related to the wrong folder name. Could you execute "yarn" on the root level and then test again?

Btw. the readme is autogenerated from the yo-generator for Theia-extensions and could potentially be outdated.

edgarmueller replied:
I see, but I suppose we can just update it? Running `yarn` on the root level works, thanks.

edgarmueller reviewed (changes requested):
Wrong folder name. If I follow 'Developing with the Browser example' section in the README I also run into issues with `Failed to resolve module: modelserver-example`. Probably related to the wrong folder name?

edgarmueller commented on modelserver-theia/package.json:
```diff
@@ -22,7 +23,10 @@
+    "@theia/core": "^0.7.1",
+    "@theia/workspace": "^0.7.1",
+    "node-fetch": "2.6.0",
+    "ip": "1.1.5"
```
Where is this package used?

edgarmueller commented on modelserver-theia/src/browser/modelserver-api.ts:
```diff
@@ -0,0 +1,82 @@
+    private restClient: RestClient;
+
+    initialize(): boolean {
+        this.modelServerBackend.getLaunchOptions().then(options => {
```
Seems as if `initialize` needs to return a `Promise<boolean>`

edgarmueller commented on modelserver-theia/src/browser/model-server-frontend-contribution.ts:
```diff
@@ -0,0 +1,36 @@
+    @inject(ModelServerApi) protected readonly modelServerService: ModelServerApi;
+
+    configure(app: FrontendApplication): MaybePromise<void> {
+        this.modelServerService.initialize();
```
If `initialize` returns a `Promise<boolean>` we should `await` the initialization here before setting up the workspace server listener

edgarmueller commented on modelserver-theia/src/browser/modelserver-api.ts:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    async get(modelUri: string): Promise<Response<string>> {
+        return this.restClient.get<string>(ModelServerPath.MODEL_CRUD + "/" + modelUri);
```
Use template literals for paths instead of concatenation.

edgarmueller commented on modelserver-theia/src/browser/rest-client.ts:
```diff
@@ -0,0 +1,97 @@
+    }
+
+
+    async  get<T>(path: string, parameters?: Map<string, string>): Promise<Response<T>> {
```
There are double spaces after `async` here and on the following methods. We should set up a formatter.

tortmayr replied:
VS Code is configured in the same fashion as GLSP (format/organize on save, ts lint active etc.)
But it seems like the default VS Code formatter does not remove redundant white spaces

edgarmueller replied:
I see, thanks for letting me know.

edgarmueller commented on example/modelsever-example/src/node/backend-module.ts:
```diff
@@ -0,0 +1,34 @@
+});
+
+@injectable()
+export class SimmpleLaunchOptions implements LaunchOptions {
```
Typo: Si**mm**ple

edgarmueller commented on example/workspace/Coffe.ecore:
```diff
@@ -0,0 +1,67 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Where do we use this file?

tortmayr replied:
I simply copied the test workspace from the modelserver repo.
I don't think it is used (yet)

edgarmueller replied:
Ok, I just ask because it seems as if the file is outdated (should be `Coffee.ecore`).

edgarmueller commented on modelserver-theia/src/common/model-server-paths.ts:
```diff
@@ -0,0 +1,26 @@
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+export namespace ModelServerPath {
```
Use plural Path**s**

edgarmueller commented on modelserver-theia/src/browser/modelserver-api.ts:
```diff
@@ -0,0 +1,82 @@
+        return this.restClient.get<string>(ModelServerPath.MODEL_CRUD + "/" + modelUri);
+    }
+
+    async  getAll(): Promise<Response<string[]>> {
```
Double spaces after `async`.

edgarmueller commented on modelserver-theia/src/browser/modelserver-api.ts:
```diff
@@ -0,0 +1,82 @@
+
+export interface ServerConfiguration {
+    workspaceRoot: string;
+    serverPort?: number;
```
Where is `serverPort` actually used?

tortmayr replied:
The serverport is as launch arg when starting the modelserver from a jar:

https://github.com/eclipsesource/modelserver-theia/blob/9e633eb599cf7f2c736465156fa8cd7371652541/modelserver-theia/src/node/model-server-backend-contribution.ts#L57

tortmayr commented:
Thanks for the review. All issues should be resolved now

edgarmueller commented on modelserver-theia/src/browser/model-server-frontend-contribution.ts:
```diff
@@ -26,11 +26,18 @@ export class ModelServerFrontendContribution implements FrontendApplicationContr
+                }
+            });
+        }
+        Promise.resolve();
```
This line has no effect.

tortmayr committed:
Address review comments

+ Update outdated example
+ Remove obsolete code line
+ Remove unused serverport argument from configuration

edgarmueller reviewed (approved):
Thank you!
